docs/core_docs/docs/tutorials/rag.ipynb

```diff
@@ -1,831 +1,1202 @@
+    "Agents](https://lilianweng.github.io/posts/2023-06-23-agent/) blog post\n",
+    "by Lilian Weng, which allows us to ask questions about the contents of\n",
+    "the post.\n",
+    "\n",
+    "We can create a simple indexing pipeline and RAG chain to do this in ~50\n",
+    "lines of code.\n",
+    "\n",
+    "```javascript\n",
+    "import \"cheerio\";\n",
+    "import { CheerioWebBaseLoader } from \"@langchain/community/document_loaders/web/cheerio\";\n",
+    "import { Document } from \"@langchain/core/documents\";\n",
+    "import { ChatPromptTemplate } from \"@langchain/core/prompts\";\n",
+    "import { pull } from \"langchain/hub\";\n",
+    "import { Annotation, StateGraph } from \"@langchain/langgraph\";\n",
+    "import { RecursiveCharacterTextSplitter } from \"@langchain/textsplitters\";\n",
+    "\n",
+    "\n",
+    "// Load and chunk contents of blog\n",
+    "const pTagSelector = \"p\";\n",
+    "const cheerioLoader = new CheerioWebBaseLoader(\n",
+    "  \"https://lilianweng.github.io/posts/2023-06-23-agent/\",\n",
+    "  {\n",
+    "    selector: pTagSelector\n",
+    "  }\n",
+    ");\n",
+    "\n",
+    "const docs = await cheerioLoader.load();\n",
+    "\n",
+    "const splitter = new RecursiveCharacterTextSplitter({\n",
+    "  chunkSize: 1000, chunkOverlap: 200\n",
+    "});\n",
+    "const allSplits = await splitter.splitDocuments(docs);\n",
+    "\n",
+    "\n",
+    "// Index chunks\n",
+    "await vectorStore.addDocuments(allSplits)\n",
+    "\n",
+    "// Define prompt for question-answering\n",
+    "const promptTemplate = await pull<ChatPromptTemplate>(\"rlm/rag-prompt\");\n",
+    "\n",
+    "// Define state for application\n",
+    "const InputStateAnnotation = Annotation.Root({\n",
+    "  question: Annotation<string>,\n",
+    "});\n",
+    "\n",
+    "const StateAnnotation = Annotation.Root({\n",
+    "  question: Annotation<string>,\n",
+    "  context: Annotation<Document[]>,\n",
+    "  answer: Annotation<string>,\n",
+    "});\n",
+    "\n",
+    "// Define application steps\n",
+    "const retrieve = async (state: typeof InputStateAnnotation.State) => {\n",
+    "  const retrievedDocs = await vectorStore.similaritySearch(state.question)\n",
+    "  return { context: retrievedDocs };\n",
+    "};\n",
+    "\n",
+    "\n",
+    "const generate = async (state: typeof StateAnnotation.State) => {\n",
+    "  const docsContent = state.context.map(doc => doc.pageContent).join(\"\\n\");\n",
+    "  const messages = await promptTemplate.invoke({ question: state.question, context: docsContent });\n",
+    "  const response = await llm.invoke(messages);\n",
+    "  return { answer: response.content };\n",
+    "};\n",
+    "\n",
+    "\n",
+    "// Compile application and test\n",
+    "const graph = new StateGraph({\n",
+    "  input: InputStateAnnotation,\n",
+    "  output: StateAnnotation,\n",
```

Comment:
I don't think you need all three of these as `StateAnnotation`

Should we just start with a single state annotation and have this as "advanced functionality"?